src/credentials/GenerateChipX509Cert.cpp

```diff
@@ -0,0 +1,464 @@
+            {
+                ReturnErrorOnFailure(EncodeCASpecificExtensions(writer));
+            }
+            else
+            {
+                ReturnErrorOnFailure(EncodeNOCSpecificExtensions(writer));
+            }
+
+            ReturnErrorOnFailure(EncodeSubjectKeyIdentifierExtension(SKI, writer));
+
+            ReturnErrorOnFailure(EncodeAuthorityKeyIdentifierExtension(AKI, writer));
+        }
+        ASN1_END_SEQUENCE;
+    }
+    ASN1_END_CONSTRUCTED;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EncodeChipDNs(ChipDNParams * params, uint8_t numParams, ASN1Writer & writer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ASN1_START_SEQUENCE
+    {
+        for (uint8_t i = 0; i < numParams; i++)
+        {
+            ASN1_START_SET
+            {
+                uint8_t chipAttrStr[kChip64bitAttrUTF8Length + 1];
+                snprintf(reinterpret_cast<char *>(chipAttrStr), sizeof(chipAttrStr), "%016" PRIX64, params[i].Value);
+
+                ASN1_START_SEQUENCE
+                {
+                    ASN1_ENCODE_OBJECT_ID(params[i].AttrOID);
+                    ReturnErrorOnFailure(writer.PutString(kASN1UniversalTag_UTF8String, Uint8::to_const_char(chipAttrStr), 16));
+                }
+                ASN1_END_SEQUENCE;
+            }
+            ASN1_END_SET;
+        }
+    }
+    ASN1_END_SEQUENCE;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EncodeValidity(uint32_t validityStart, uint32_t validityEnd, ASN1Writer & writer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    ASN1UniversalTime asn1Time;
+
+    ASN1_START_SEQUENCE
+    {
+        ReturnErrorOnFailure(ChipEpochToASN1Time(validityStart, asn1Time));
+        ASN1_ENCODE_TIME(asn1Time);
+
+        ReturnErrorOnFailure(ChipEpochToASN1Time(validityEnd, asn1Time));
+        ASN1_ENCODE_TIME(asn1Time);
+    }
+    ASN1_END_SEQUENCE;
+
+exit:
+    return err;
+}
+
+CHIP_ERROR EncodeChipECDSASignature(Crypto::P256ECDSASignature & signature, ASN1Writer & writer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    ASN1_START_BIT_STRING_ENCAPSULATED { writer.PutConstructedType(signature, (uint16_t) signature.Length()); }
+    ASN1_END_ENCAPSULATED;
+
+exit:
+    return err;
+}
+
+} // namespace
+
+CHIP_ERROR EncodeTBSCert(const X509CertRequestParams & requestParams, CertificateIssuerLevel issuerLevel, uint64_t subject,
+                         const Crypto::P256PublicKey & subjectPubkey, const Crypto::P256PublicKey & issuerPubkey,
+                         ASN1Writer & writer)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    ChipDNParams dnParams[2];
+    uint8_t numDNs = 1;
+    bool isCA      = true;
+
+    ASN1_START_SEQUENCE
+    {
+        // version [0] EXPLICIT Version DEFAULT v1
+        ASN1_START_CONSTRUCTED(kASN1TagClass_ContextSpecific, 0)
+        {
+            // Version ::= INTEGER { v1(0), v2(1), v3(2) }
+            ASN1_ENCODE_INTEGER(2);
+        }
+        ASN1_END_CONSTRUCTED;
+
+        ReturnErrorOnFailure(writer.PutInteger(requestParams.SerialNumber));
```

Comment:
Must verify that `requestParams.SerialNumber` is a positive integer.

Reply:
Added a check, and tests.